source/cbor.c

```diff
@@ -0,0 +1,739 @@
+                AWS_ASSERT(decoder->error_code != AWS_ERROR_SUCCESS);
+                break;
+            }
+            /* convert the value to -1 * expected to do the math as unsigned. */
+            timestamp_secs = aws_add_u64_saturating(timestamp_secs, 1);
+            uint64_t timestamp_ms = aws_mul_u64_saturating(timestamp_secs, 1000);
+            if (timestamp_ms > INT64_MAX) {
+                goto overflow;
+            }
+            *out = (int64_t)timestamp_ms * -1LL;
+            break;
+        }
+
+        case AWS_CBOR_TYPE_FLOAT: {
+            double timestamp_secs = 0;
+            error |= aws_cbor_decoder_pop_next_float_val(decoder, &timestamp_secs);
+            if (error) {
+                /* The error code for decoder must have been set as we already checked the type. */
+                AWS_ASSERT(decoder->error_code != AWS_ERROR_SUCCESS);
+                break;
+            }
+            /* TODO: How to check overflow? */
+            double ms_double = timestamp_secs * 1000.0;
+            /* Round it up instead of type cast incase of overflow */
+            int64_t timestamp_ms = llround(ms_double);
+            if (fetestexcept(FE_INVALID) || timestamp_ms > INT64_MAX) {
+                goto overflow;
+            }
+            *out = timestamp_ms;
+            break;
+        }
+        default:
+            AWS_LOGF_ERROR(AWS_LS_COMMON_CBOR, "The cbor data is malformed to decode.");
+            decoder->error_code = AWS_ERROR_INVALID_CBOR;
+            error = aws_raise_error(AWS_ERROR_INVALID_CBOR);
+            break;
+    }
+    if (error) {
+        return AWS_OP_ERR;
+    }
+    return AWS_OP_SUCCESS;
+overflow:
+    /* Overflow detected, it's technically not a malformed cbor item, but for this helper */
+    AWS_LOGF_ERROR(AWS_LS_COMMON_CBOR, "The decoded timestamp cannot fit into int64_t.");
+    /* TODO: do we want to revert the decoded part???? Probably??? */
+    return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+}
+
+int aws_cbor_decoder_consume_next_data_item(struct aws_cbor_decoder *decoder) {
+
+    if (decoder->cached_context.type == AWS_CBOR_TYPE_MAX) {
+        /* There was no cache, decode the next item */
+        if (s_cbor_decode_next_element(decoder)) {
+            return AWS_OP_ERR;
+        }
+    }
+    switch (decoder->cached_context.type) {
+        case AWS_CBOR_TYPE_TAG:
+            /* Read the next data item */
+            /* TODO: error check for the tag content?? */
+            decoder->cached_context.type = AWS_CBOR_TYPE_MAX;
+            if (aws_cbor_decoder_consume_next_data_item(decoder)) {
+                return AWS_OP_ERR;
+            }
+            break;
+        case AWS_CBOR_TYPE_MAP_START: {
+            uint64_t num_map_item = decoder->cached_context.cbor_data.map_start;
+            /* Reset type */
+            decoder->cached_context.type = AWS_CBOR_TYPE_MAX;
+            for (uint64_t i = 0; i < num_map_item; i++) {
```

Comment:
Could this be a vector for some kind of attack? Like, attacker sends CBOR with a map or array that says it's UINT64_MAX items. Do we pre-allocate space anywhere, that could lead to a out-of-memory crash? Or do we loop through all items?

Or it's not really possible to attack that way on decode? Because we just decode data as it comes in, and they'd actually have to send all that data (not just claim they were GOING to send all that data) ???